packages/caravan-fees/src/utils.ts

```diff
@@ -0,0 +1,839 @@
+    case Network.MAINNET:
+      return networks.bitcoin;
+    case Network.TESTNET:
+      return networks.testnet;
+    case Network.REGTEST:
+      return networks.regtest;
+    case Network.SIGNET:
+      // As of the last check, bitcoinjs-lib doesn't have built-in support for signet.
+      // If signet support is crucial, you might need to define a custom network.
+      throw new Error(
+        "Signet is not directly supported in bitcoinjs-lib. Consider defining a custom network if needed.",
+      );
+    default:
+      throw new Error(`Unsupported network: ${network}`);
+  }
+}
+
+/**
+ * Estimates the virtual size (vsize) of a transaction represented by a PsbtV2 object.
+ *
+ * This function provides a comprehensive estimation of the transaction's vsize,
+ * which is crucial for fee calculation in Bitcoin transactions. It supports
+ * various output types including P2PKH, P2SH, P2WPKH, and P2WSH, as well as
+ * their nested variants (e.g., P2SH-P2WPKH, P2SH-P2WSH).
+ *
+ * The estimation takes into account:
+ * - Transaction version (4 bytes)
+ * - Input and output counts (1-9 bytes each, depending on the count)
+ * - Input details (outpoints, scriptSigs, witnesses)
+ * - Output details (value and scriptPubKey)
+ * - Locktime (4 bytes)
+ *
+ * For SegWit transactions, it calculates the weight units and converts to vsize.
+ *
+ * References:
+ * - BIP141 (SegWit): https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki
+ * - BIP143 (Transaction Signature Verification for Version 0 Witness Program):
+ *   https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki
+ * - Bitcoin Core weight calculation:
+ *   https://github.com/bitcoin/bitcoin/blob/master/src/consensus/validation.h
+ *
+ * @param {PsbtV2} psbt - The PsbtV2 object representing the transaction.
+ * @returns {number} The estimated virtual size (vsize) of the transaction in vbytes.
+ *
+ * @example
+ * const psbt = new PsbtV2(...);
+ * const estimatedVsize = estimatePSBTvsize(psbt);
+ * console.log(`Estimated vsize: ${estimatedVsize} vbytes`);
+ */
+export function estimatePSBTvsize(psbt: PsbtV2): number {
+  let weight = 0;
+  let hasWitness = false;
+
+  // Transaction version (4 bytes)
+  weight += 4 * 4;
+
+  // Input count
+  weight += getVarIntSize(psbt.PSBT_GLOBAL_INPUT_COUNT) * 4;
+
+  // Inputs
+  for (let i = 0; i < psbt.PSBT_GLOBAL_INPUT_COUNT; i++) {
+    weight += estimateInputWeight(psbt, i);
+    if (psbt.PSBT_IN_WITNESS_UTXO[i] || psbt.PSBT_IN_WITNESS_SCRIPT[i]) {
+      hasWitness = true;
+    }
+  }
+
+  // Output count
+  weight += getVarIntSize(psbt.PSBT_GLOBAL_OUTPUT_COUNT) * 4;
+
+  // Outputs
+  for (let i = 0; i < psbt.PSBT_GLOBAL_OUTPUT_COUNT; i++) {
+    weight += estimateOutputWeight(psbt, i);
+  }
+
+  // Locktime (4 bytes)
+  weight += 4 * 4;
+
+  // If it's a SegWit transaction, add marker and flag (2 bytes)
+  if (hasWitness) {
+    weight += 2;
+  }
+
+  // Convert weight to vsize (rounded up)
+  return Math.ceil(weight / 4);
```

Comment:
>  I thought you had a size calculator for transactions already? 

the size calculator you are referencing `estimateTransactionVsize` is not for calculating the vsize of a PSBT so thta's why we have defined `estimatePSBTvsize` fn that takes in a PSBT object ...

Reply:
So maybe we can use the psbt utility from bitcoin-js itself ?

Reply:
Why would the size calculation of the psbt and tx be different? I would’ve expected a psbt calculation to not actually calculate the size of the psbt (since that’s not really relevant to much except maybe network bandwidth constraints). But yeah the bitcoinjs utility should work but I imagine it’s doing the same thing (estimating the size of the underlying tx data)

Reply:
So, you're totally right. theoretically you could have a transaction with inputs of different types. But that's going to be basically a coinjoin where you're taking utxos from multiple different wallets. In practice this is very rare, and not supported in Caravan Coordinator of course which assumes spending from a single wallet.

Outputs definitely can and will though and does need to take that into account.

Reply:
Removed this utility now